transform.c:59:23: error: implicit declaration of function ‘strchr’

Dear developers,


Many thanks for developing and maintaining your tools all these years.
Apologies if this error was reported before, I am quite new to bug reporting on github and still learning this area.

I am building MEEP and MPB from sources on Ubuntu 26.04 LTS with all the latest updates applied. 
This error appears during the compilation of MEEP/MPB code:

->
make[3]: Entering directory '/home/pavlo/MEEP_Install/mpb/mpb'
  CC       mpb_mpi-transform.o
transform.c: In function ‘transformed_overlap’:
transform.c:59:23: error: implicit declaration of function ‘strchr’ [-Wimplicit-function-declaration]
   59 |     if (!curfield || !strchr("db", curfield_type)) {
      |                       ^~~~~~
transform.c:31:1: note: include ‘<string.h>’ or provide a declaration of ‘strchr’
   30 | #include <xyz_loop.h>
  +++ |+#include <string.h>
   31 | #include <maxwell.h>
transform.c:59:23: warning: incompatible implicit declaration of built-in function ‘strchr’ [-Wbuiltin-declaration-mismatch]
   59 |     if (!curfield || !strchr("db", curfield_type)) {
      |                       ^~~~~~
transform.c:59:23: note: include ‘<string.h>’ or provide a declaration of ‘strchr’
<-

I do not know how severe it is (will see latter today) but decided to report it any way.

best wishes,
pavlo

The error message thrown out in the process of MPB compilation.

eigs_test.c:350:23: warning: unused parameter ‘Y’ [-Wunused-parameter]
  350 |           evectmatrix Y, real *eigenvals, sqmatrix YtY)
      |           ~~~~~~~~~~~~^
  CCLD     eigs_test
  CC       maxwell_test.o
  CCLD     maxwell_test
  CC       normal_vectors-normal_vectors.o
  CCLD     normal_vectors
make[2]: Leaving directory '/home/pavlo/MEEP_Install/mpb/tests'
Making all in mpb
make[2]: Entering directory '/home/pavlo/MEEP_Install/mpb/mpb'
cp -f /usr/local/share/libctl/base/main.c main.c
cp -f /usr/local/share/libctl/utils/geom.c geom.c
gen-ctl-io --header -o ctl-io.h mpb.scm /usr/local/share/libctl
gen-ctl-io --code -o ctl-io.c mpb.scm /usr/local/share/libctl
make  all-am
make[3]: Entering directory '/home/pavlo/MEEP_Install/mpb/mpb'
  CC       mpb_mpi-transform.o
transform.c: In function ‘transformed_overlap’:
transform.c:59:23: error: implicit declaration of function ‘strchr’ [-Wimplicit-function-declaration]
   59 |     if (!curfield || !strchr("db", curfield_type)) {
      |                       ^~~~~~
transform.c:31:1: note: include ‘<string.h>’ or provide a declaration of ‘strchr’
   30 | #include <xyz_loop.h>
  +++ |+#include <string.h>
   31 | #include <maxwell.h>
transform.c:59:23: warning: incompatible implicit declaration of built-in function ‘strchr’ [-Wbuiltin-declaration-mismatch]
   59 |     if (!curfield || !strchr("db", curfield_type)) {
      |                       ^~~~~~
transform.c:59:23: note: include ‘<string.h>’ or provide a declaration of ‘strchr’
In file included from transform.c:30:
../src/maxwell/xyz_loop.h:28:28: warning: unused variable ‘n2’ [-Wunused-variable]
   28 |           int n1 = md->nx, n2 = md->ny, n3 = md->nz, i1, i2_, i3; \
      |                            ^~
transform.c:118:5: note: in expansion of macro ‘LOOP_XYZ’
  118 |     LOOP_XYZ(mdata) { /* implies two opening braces `{{` */
      |     ^~~~~~~~
make[3]: *** [Makefile:600: mpb_mpi-transform.o] Error 1
make[3]: Leaving directory '/home/pavlo/MEEP_Install/mpb/mpb'
make[2]: *** [Makefile:432: all] Error 2
make[2]: Leaving directory '/home/pavlo/MEEP_Install/mpb/mpb'
make[1]: *** [Makefile:437: all-recursive] Error 1
make[1]: Leaving directory '/home/pavlo/MEEP_Install/mpb'
make: *** [Makefile:369: all] Error 2
python3-h5py is already the newest version (3.15.1-4build1).
Summary:                    
  Upgrading: 0, Installing: 0, Removing: 0, Not Upgrading: 0
python3-debugpy is already the newest version (1.8.19+ds-1ubuntu3).
Summary:                    
  Upgrading: 0, Installing: 0, Removing: 0, Not Upgrading: 0

--- a/mpb/fields.c
+++ b/mpb/fields.c
@@ -17,6 +17,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <math.h>
 #include <ctype.h>
 #include <stddef.h>

--- a/mpb/material_grid_opt.c
+++ b/mpb/material_grid_opt.c
@@ -63,7 +63,7 @@ static double mindiff_func(int n, const double *u, double *grad, void *data)
     solve_kpoint(d->k);
     gap = (f2 = freqs.items[d->b-1]);
     if (grad) {
-        material_grids_addgradient(work, 1.0, d->b, 
+        material_grids_addgradient(work, 1.0, d->b,
                                    d->grids, d->ngrids);
     }
 
@@ -72,15 +72,15 @@ static double mindiff_func(int n, const double *u, double *grad, void *data)
     solve_kpoint(d->k);
     gap -= (f1 = freqs.items[d->b-1]);
     if (grad) {
-        material_grids_addgradient(work, -1.0, d->b, 
+        material_grids_addgradient(work, -1.0, d->b,
                                    d->grids, d->ngrids);
     }
 
     if (grad) /* gradient w.r.t. epsilon needs to be summed over processes */
-        mpi_allreduce(work, grad, n, double, MPI_DOUBLE, 
+        mpi_allreduce(work, grad, n, double, MPI_DOUBLE,
                       MPI_SUM, mpb_comm);
 
-     mpi_one_printf("material-grid-mindiff:, %d, %g, %g, %0.15g\n", 
+     mpi_one_printf("material-grid-mindiff:, %d, %g, %g, %0.15g\n",
                     d->iter, f1, f2, gap);
 
      return gap;
@@ -191,7 +191,7 @@ static double band_constraint(int n, const double *u, double *grad, void *data)
      if (grad) memset(work, 0, sizeof(double) * (n-2));
      if (kind == BAND1_CONSTRAINT) {
 	  if (grad) {
-	       material_grids_addgradient(work, 1.0, d->b1, 
+	       material_grids_addgradient(work, 1.0, d->b1,
 					  d->grids, d->ngrids);
 	       grad[n-1] = -1;
 	       grad[n-2] = 0;
@@ -200,7 +200,7 @@ static double band_constraint(int n, const double *u, double *grad, void *data)
      }
      else {
 	  if (grad) {
-	       material_grids_addgradient(work, -1.0, d->b2, 
+	       material_grids_addgradient(work, -1.0, d->b2,
 					  d->grids, d->ngrids);
 	       grad[n-1] = 0;
 	       grad[n-2] = 1;
@@ -208,7 +208,7 @@ static double band_constraint(int n, const double *u, double *grad, void *data)
 	  val = u[n-2] - (d->f2s[ik] = freqs.items[d->b2-1]);
      }
      if (grad) /* gradient w.r.t. epsilon needs to be summed over processes */
-	  mpi_allreduce(work, grad, n-2, double, MPI_DOUBLE, 
+	  mpi_allreduce(work, grad, n-2, double, MPI_DOUBLE,
 			MPI_SUM, mpb_comm);
 
      return val;
@@ -227,7 +227,7 @@ static double maxgap_func(int n, const double *u, double *grad, void *data)
      d->unsolved = 1;
 
      gap = (f2 - f1) * 2.0 / (f1 + f2);
-     
+
      if (grad) {
 	  memset(grad, 0, sizeof(double) * (n-2));
 	  grad[n-1] = 2.0 * ((f1 + f2) - (f1 - f2)) / ((f1+f2)*(f1+f2));
@@ -238,13 +238,13 @@ static double maxgap_func(int n, const double *u, double *grad, void *data)
 	  }
      }
 
-     mpi_one_printf("material-grid-%sgap:, %d, %g, %g, %0.15g\n", 
+     mpi_one_printf("material-grid-%sgap:, %d, %g, %g, %0.15g\n",
 		    d->do_min ? "min" : "max", d->iter, f1, f2, gap);
-     
+
      if (verbose) {
 	  char prefix[256];
 	  get_epsilon();
-	  snprintf(prefix, 256, "%sgap-%04d-", 
+	  snprintf(prefix, 256, "%sgap-%04d-",
 		   d->do_min ? "min" : "max", d->iter);
 	  output_field_to_file(-1, prefix);
      }
@@ -253,7 +253,7 @@ static double maxgap_func(int n, const double *u, double *grad, void *data)
 }
 
 static number material_grids_maxmin_gap(boolean do_min,
-					vector3_list kpoints, 
+					vector3_list kpoints,
 					integer band1, integer band2,
 					number func_tol, number eps_tol,
 					integer maxeval, number maxtime)
@@ -323,7 +323,7 @@ static number material_grids_maxmin_gap(boolean do_min,
      mma_verbose = kpoints.num_items*2;
      res = nlopt_minimize_constrained(
 	  NLOPT_LD_MMA, n, maxgap_func, &d,
-	  kpoints.num_items*2, band_constraint, 
+	  kpoints.num_items*2, band_constraint,
 	  cdata, sizeof(band_constraint_data),
 	  lb, ub, u, &func_min,
 	  -HUGE_VAL, func_tol,0, 0,u_tol, maxeval,maxtime);
@@ -352,7 +352,7 @@ static number material_grids_maxmin_gap(boolean do_min,
 
      func_min = (u[n-2] - u[n-1]) * 2.0 / (u[n-1] + u[n-2]);
      mpi_one_printf("material-grid-%sgap:, %d, %g, %g, %0.15g\n",
-                    d.do_min ? "min" : "max", d.iter+1, 
+                    d.do_min ? "min" : "max", d.iter+1,
 		    u[n-1], u[n-2], func_min);
      func_min = d.do_min ? func_min : -func_min;
 
@@ -364,7 +364,7 @@ static number material_grids_maxmin_gap(boolean do_min,
      return(do_min ? func_min : -func_min);
 }
 
-number material_grids_maxgap(vector3_list kpoints, 
+number material_grids_maxgap(vector3_list kpoints,
 			     integer band1, integer band2,
 			     number func_tol, number eps_tol,
 			     integer maxeval, number maxtime)
@@ -373,7 +373,7 @@ number material_grids_maxgap(vector3_list kpoints,
 				      func_tol, eps_tol, maxeval, maxtime);
 }
 
-number material_grids_mingap(vector3_list kpoints, 
+number material_grids_mingap(vector3_list kpoints,
 			     integer band1, integer band2,
 			     number func_tol, number eps_tol,
 			     integer maxeval, number maxtime)

--- a/mpb/mpb.c
+++ b/mpb/mpb.c
@@ -194,7 +194,7 @@ number mpi_max(number num)
 /**************************************************************************/
 /* expose some build info to guile */
 
-boolean has_hermitian_epsp()
+boolean has_hermitian_epsp(void)
 {
 #if WITH_HERMITIAN_EPSILON
     return 1;
@@ -203,7 +203,7 @@ boolean has_hermitian_epsp()
 #endif
 }
 
-boolean has_inversion_symp()
+boolean has_inversion_symp(void)
 {
 #if SCALAR_COMPLEX
     return 0;
@@ -869,7 +869,7 @@ number_list compute_yparities(void)
    group velocities, one for each band, in units of c. */
 number_list compute_group_velocity_component(vector3 d)
 {
-     number_list group_v;
+     number_list group_v = { 0, (number *) NULL };
      real *gv_scratch;
      real u[3];
      int i, ib;
@@ -951,7 +951,7 @@ number compute_1_group_velocity_component(vector3 d, integer b)
 {
      real u[3];
      int ib = b - 1;
-     real group_v, scratch;
+     real group_v = 0, scratch;
 
      curfield_reset();
 

--- a/mpb/mpb.h
+++ b/mpb/mpb.h
@@ -111,5 +111,7 @@ void material_grids_addgradient(double *v,
 /**************************************************************************/
 
 extern const char *parity_string(maxwell_data *d);
+boolean has_hermitian_epsp(void);
+boolean has_inversion_symp(void);
 
 #endif /* MPB_H */

--- a/mpb/transform.c
+++ b/mpb/transform.c
@@ -18,6 +18,7 @@
 
 #include <stdio.h>
 #include <stdlib.h>
+#include <string.h>
 #include <math.h>
 
 #include "config.h"
@@ -32,9 +33,9 @@
 
 
 
-/* If `curfield` is the real-space D-field (of band-index i), computes the overlap 
+/* If `curfield` is the real-space D-field (of band-index i), computes the overlap
  *       ∫ Eᵢ†(r){W|w}Dᵢ(r) dr  =  ∫ Eᵢ†(r)(WDᵢ)({W|w}⁻¹r) dr,
- * for a symmetry operation {W|w} with rotation W and translation w; each specified 
+ * for a symmetry operation {W|w} with rotation W and translation w; each specified
  * in the lattice basis. The vector fields Eᵢ and Dᵢ include Bloch phases.
  * If `curfield` is the real-space B-field, the overlap
  *       ∫ Hᵢ†(r){W|w}Bᵢ(r) dr  =  det(W) ∫ Hᵢ†(r)(WBᵢ)({W|w}⁻¹r) dr,
@@ -43,7 +44,7 @@
  * independent of whether the D- or B-field is used.
  * No other choices for `curfield` are allowed: to set `curfield` to the real-space
  * B- or D-field use the `get-bfield` and `get-dfield` Scheme functions or the
- * `get_bfield` and `get_dfield` in C functions.                   
+ * `get_bfield` and `get_dfield` in C functions.
  * Usually, it will be more convenient to use the wrapper `compute_symmetry(i, W, w)`
  * which defaults to the B-field (since μ = 1 usually) and takes a band-index `i`.
  */
@@ -70,7 +71,7 @@ cnumber transformed_overlap(matrix3x3 W, vector3 w)
          * counting the number of grid points correctly somehow: I think the root issue is
          * that we use the LOOP_XYZ macro, which has special handling for mbpi (i.e. only
          * "visits" the "inversion-reduced" part of the unitcell): but here, we actually
-         * really want to (or at least assume to) visit _all_ the points in the unitcell. 
+         * really want to (or at least assume to) visit _all_ the points in the unitcell.
          */
     #endif
     #ifdef HAVE_MPI
@@ -83,7 +84,7 @@ cnumber transformed_overlap(matrix3x3 W, vector3 w)
          * Long story short: disable it for now.
          */
     #endif
-    
+
     /* prepare before looping ... */
     n1 = mdata->nx; n2 = mdata->ny; n3 = mdata->nz;
 
@@ -128,7 +129,7 @@ cnumber transformed_overlap(matrix3x3 W, vector3 w)
         /* transformed coordinate pt = {W|w}⁻¹p = W⁻¹(p-w) since {W|w}⁻¹={W⁻¹|-W⁻¹w} */
         pt = matrix3x3_vector3_mult(invW, vector3_minus(p, w));
 
-        /* Bloch field value at transformed coordinate pt: interpolation is needed to 
+        /* Bloch field value at transformed coordinate pt: interpolation is needed to
          * ensure generality in the case of fractional translations.                  */
         get_bloch_field_point_(Ftemp, pt); /* assign `Ftemp` to field at `pt` (without eⁱᵏʳ factor) */
 
@@ -157,7 +158,7 @@ cnumber transformed_overlap(matrix3x3 W, vector3 w)
             F[1] = curfield[3*xyz_index+1];
             F[2] = curfield[3*xyz_index+2];
         }
-        
+
         /* inner product of F and Ft={W|w}F in Bloch form */
         CASSIGN_CONJ_MULT(integrand, F[0], Ft[0]);  /* add adjoint(F)*Ft to integrand */
         CACCUMULATE_SUM_CONJ_MULT(integrand, F[1], Ft[1]);

